src/neo/Network/P2P/TaskManager.cs

```diff
@@ -274,56 +231,60 @@ public static Props Props(NeoSystem system)
-            var highestBlockIndex = sessions.Values.Max(p => p.LastBlockIndex);
-            for (; taskCounts < MaxSyncTasksCount; taskCounts++)
+            uint currentHeight = NativeContract.Ledger.CurrentIndex(snapshot);
+            uint headerHeight = Blockchain.Singleton.HeaderCache.Last?.Index ?? currentHeight;
+            // When the number of AvailableTasks is no more than 0, no pending tasks of InventoryType.Block, it should process pending the tasks of headers
+            // If not HeaderTask pending to be processed it should ask for more Blocks
+            if ((!HasHeaderTask || globalTasks[HeaderTaskHash] < MaxConncurrentTasks) && headerHeight < session.LastBlockIndex && !Blockchain.Singleton.HeaderCache.Full)
             {
-                if (lastTaskIndex >= highestBlockIndex || lastTaskIndex >= currentHeight + InvPayload.MaxHashesCount) break;
-                if (!AssignSyncTask(++lastTaskIndex)) break;
+                session.Tasks[HeaderTaskHash] = DateTime.UtcNow;
+                IncrementGlobalTask(HeaderTaskHash);
+                remoteNode.Tell(Message.Create(MessageCommand.GetHeaders, GetBlockByIndexPayload.Create(headerHeight)));
             }
-        }
-
-        private void SendPingMessage()
-        {
-            DataCache snapshot = Blockchain.Singleton.View;
-            uint currentHeight = NativeContract.Ledger.CurrentIndex(snapshot);
-            UInt256 currentHash = NativeContract.Ledger.CurrentHash(snapshot);
-            TrimmedBlock block = NativeContract.Ledger.GetTrimmedBlock(snapshot, currentHash);
-            foreach (KeyValuePair<IActorRef, TaskSession> item in sessions)
+            else if (currentHeight < session.LastBlockIndex)
             {
-                var node = item.Key;
-                var session = item.Value;
-
-                if (session.ExpireTime < TimeProvider.Current.UtcNow ||
-                     (block.Index >= session.LastBlockIndex &&
-                     TimeProvider.Current.UtcNow.ToTimestampMS() - PingCoolingOffPeriod >= block.Timestamp))
+                uint startHeight = currentHeight;
+                foreach (Header header in Blockchain.Singleton.HeaderCache)
                 {
-                    if (session.InvTasks.Remove(MemPoolTaskHash))
-                    {
-                        node.Tell(Message.Create(MessageCommand.Mempool));
-                    }
-                    node.Tell(Message.Create(MessageCommand.Ping, PingPayload.Create(currentHeight)));
-                    session.ExpireTime = TimeProvider.Current.UtcNow.AddMilliseconds(PingCoolingOffPeriod);
+                    if (header.Index <= currentHeight) continue;
+                    if (!globalTasks.ContainsKey(header.Hash)) break;
+                    ++startHeight;
                 }
+                remoteNode.Tell(Message.Create(MessageCommand.GetBlockByIndex, GetBlockByIndexPayload.Create(startHeight)));
```

Comment:
It might create some repetitive tasks here.

Reply:
Yes. That's why we need to recover the `IndexTasks`.